### ISSUE TITLE
Calculation of the loss

### ISSUE THREAD

**Issue by JesseYang** [ibab/tensorflow-wavenet#98]
In each step, the calculation of the loss is based on the cross entropy of the input value and the predicted value. Each data point in the predicted value is predicted based on its receptive field in the input value. During prediction, the input value is padded on the left with many 0s to implement the casual convolution. Thus, the beginning predicted data points are actually predicted based on those padding 0s, and may not be the same as the input values even when the model is trained for many steps. Is this the reason that why the loss drops to around 2 and cannot drop lower? Maybe it is more reasonable to calculate the loss not from the beginning data point, but from the N-th point where N is the receptive field.

**Comment by nakosung** [comment]
If so, final output layer should be truncated not to report false losses? I think it might do harm with network parameters.

**Comment by nakosung** [comment]
Implementation in other repo: https://github.com/basveeling/wavenet/blob/master/wavenet.py#L491

**Comment by jyegerlehner** [comment]
@JesseYang 
OK I think I finally understand what you are saying. I think you are right. I'll try to put it in my own words: when computing the discrete prob distribution for the next sample, the net needs (and usually has) all the prior values available to it in order to compute that next predicted sample. The exception is: any time in the time history less than N samples from the start (where N is the receptive field width), we are penalizing the net for failing to predict the output, when we haven't given it all the inputs it needs. In fact, the first sample it predicts has no input values at all available to it.  Yeah. 

So the concern is not that the displayed loss number is a bit inflated. Rather, that we back-propagate errors and adjust net weights and biases for cases where we haven't given the net all the inputs it needs. That seems it would have a distorting, corrupting effect.

Most clips are several seconds long, and the net may have a, say, 300 msec receptive field. So it's a relatively small portion of the weight updates that have this corruption, if it can be called that. 

@nakosung Is your code there a proposed modification of model.py loss method, whereby we exclude the first N samples from the loss calculation? If so I think I agree that would be a fix. Though I'm not familiar with tf.slice and can't judge its correctness yet.

**Comment by nakosung** [comment]
@jyegerlehner Most fast way would be using masking out first N samples from calculated loss just before reducung it.

**Comment by akademi4eg** [comment]
Seems like this is resolved now with merge of PR 191.

**Comment by Fjolnir-Dvorak** [comment]
It actually appeared yesterday again in a training session on the newest commit. I will try to pin it down. The only problem: I trained without metadata and log. I will add a proof of failure this week.
